src/timeline/TimelineModel.cpp

```diff
@@ -1471,6 +1474,107 @@ TimelineModel::formatTypingUsers(const std::vector<QString> &users, QColor bg)
+        case mtx::events::state::JoinRule::Invite:
+                return tr("%1 closed the room to be by invitation only").arg(name);
+        default:
+                // Currently, knock and private are reserved keywords and not implemented in Matrix.
+                return "";
+        }
+}
+
+QString
+TimelineModel::formatGuestAccessEvent(QString id)
+{
+        if (!events.contains(id))
+                return "";
+
+        auto event =
+          std::get_if<mtx::events::StateEvent<mtx::events::state::GuestAccess>>(&events[id]);
+        if (!event)
+                return "";
+
+        QString user = QString::fromStdString(event->sender);
+        QString name = escapeEmoji(displayName(user));
+
+        switch (event->content.guest_access) {
+        case mtx::events::state::AccessState::CanJoin:
+                return tr("%1 made the room open to guests").arg(name);
+        case mtx::events::state::AccessState::Forbidden:
+                return tr("%1 has closed the room to guest access").arg(name);
+        default:
+                return "";
+        }
+}
+
+QString
+TimelineModel::formatHistoryVisibilityEvent(QString id)
+{
+        if (!events.contains(id))
+                return "";
+
+        auto event =
+          std::get_if<mtx::events::StateEvent<mtx::events::state::HistoryVisibility>>(&events[id]);
+
+        if (!event)
+                return "";
+
+        QString user = QString::fromStdString(event->sender);
+        QString name = escapeEmoji(displayName(user));
+
+        switch (event->content.history_visibility) {
+        case mtx::events::state::Visibility::WorldReadable:
+                return tr("%1 made the room history world readable. Events may be now read by "
+                          "non-joined people")
+                  .arg(name);
+        case mtx::events::state::Visibility::Shared:
+                return tr("%1 set the room history visible to members from this point on")
+                  .arg(name);
+        case mtx::events::state::Visibility::Invited:
+                return tr("%1 set the room history visible to members since they were invited")
+                  .arg(name);
+        case mtx::events::state::Visibility::Joined:
+                return tr("%1 set the room history visible to members since they joined the room")
+                  .arg(name);
+        default:
+                return "";
+        }
+}
+
+QString
+TimelineModel::formatPowerLevelEvent(QString id)
+{
+        if (!events.contains(id))
+                return "";
+
+        auto event =
+          std::get_if<mtx::events::StateEvent<mtx::events::state::PowerLevels>>(&events[id]);
+        if (!event)
+                return "";
+
+        QString user = QString::fromStdString(event->sender);
+        QString name = escapeEmoji(displayName(user));
+
+        // TODO: power levels rendering is actually a bit complex. work on this later.
+        return tr("%1 has changed the room's permissions.").arg(name);
```

Comment:
If you wanted to make the changes from https://github.com/Nheko-Reborn/nheko/pull/131/files#diff-635ee45863a598079569b6d91d50c8e1R1574-R1582 work with the translation, you need something like:

```suggestion
        QString* message[] = {
                              //: Comment for the translator: one of the following strings is randomly chosen on powerlevel changes, have fun!
                               tr("%1 has altered the deal, pray they do not alter it any further."),
                               tr("%1 has increased power levels to over 9000!"),
                               tr("%1 rolls for a power check, critical hit."),
                               tr("%1 has tipped the scales."),
                               tr("%1 thinks the rules need to change.")
                               };

        QString *chosen = message[rand() % sizeof(message)];

        // TODO: power levels rendering is actually a bit complex. work on this later.
        return chosen->arg(name);
```

(this is untested). I am not sure how to allow a different number of translated strings in multiple languages.

Reply:
thanks! hope thisdoesn't get lost if we decide to re-inject jokes before the power levels stuff gets fully implemented.